src/engine/input/InputManager.cs

```diff
@@ -0,0 +1,265 @@
+    }
+
+    /// <summary>
+    ///   Calls the OnInput methods of the attributes where ActivationType is Input.
+    ///   Ignores InputEventMouseMotion events.
+    ///   Sets the input as consumed, if it was consumed.
+    /// </summary>
+    /// <param name="event">The event the user fired</param>
+    public override void _Input(InputEvent @event)
+    {
+        OnInput(false, @event);
+    }
+
+    public override void _Notification(int focus)
+    {
+        // If the window goes out of focus, we don't receive the key released events
+        // We reset our held down keys if the player tabs out while pressing a key
+        if (focus == MainLoop.NotificationWmFocusOut)
+        {
+            OnFocusLost();
+        }
+    }
+
+    internal static bool CallMethod(InputAttribute attribute, object[] parameters)
+    {
+        var method = attribute.Method;
+
+        // Do nothing if no method is associated
+        // TODO: it would be safer against mistakes to make it so that only specific attributes can have null method
+        if (method == null)
+            return true;
+
+        var disposed = new List<WeakReference>();
+        var instances = staticInstance.attributes[attribute];
+        var result = false;
+
+        if (method.IsStatic)
+        {
+            // Call the method without an instance if it's static
+            var invokeResult = method.Invoke(null, parameters);
+
+            if (invokeResult is bool asBool)
+            {
+                result = asBool;
+            }
+            else
+            {
+                result = true;
+            }
+        }
+        else
+        {
+            // Call the method for each instance
+            foreach (var instance in instances)
+            {
+                if (!instance.IsAlive)
+                {
+                    // if the WeakReference got disposed
+                    disposed.Add(instance);
+                    continue;
+                }
+
+                bool thisInstanceResult;
+
+                var invokeResult = method.Invoke(instance.Target, parameters);
+
+                if (invokeResult != null && invokeResult is bool asBool)
+                {
+                    thisInstanceResult = asBool;
+                }
+                else
+                {
+                    thisInstanceResult = true;
+                }
+
+                if (thisInstanceResult)
+                {
+                    result = true;
+                }
+            }
+        }
+
+        disposed.ForEach(p => instances.Remove(p));
+
+        return result;
+    }
+
+    private void OnInput(bool inputUnhandled, InputEvent @event)
+    {
+        // Ignore mouse motion
+        // TODO: support mouse movement input as well
+        if (@event is InputEventMouseMotion)
+            return;
+
+        var handled = attributes.Any(
+            attribute => (inputUnhandled || !attribute.Key.OnlyUnhandled) && attribute.Key.OnInput(@event));
```

Comment:
I'd prefer this to be a written out loop (instead of LINQ) just because the ordering matters because of the input consuming.